libs/logger/logger.cpp

```diff
@@ -5,58 +5,262 @@
-    auto logger = spdlog::get(tag);
-    if (logger == nullptr) {
-      logger = createLogger(tag);
+  class Logger::Impl {
+   public:
+    virtual ~Impl() = default;
+
+    /// Constructor.
+    ///
+    /// @param tag - the tag of constructed logger
+    /// @param tree - the logger tree configuration
+    /// @param spdlog_logger - the spdlog logger to use
+    Impl(std::string tag,
+         ConstLoggerConfigTreeNodePtr tree,
+         std::shared_ptr<spdlog::logger> spdlog_logger)
+        : tag_(std::move(tag)),
+          tree_(std::move(tree)),
+          logger_(std::move(spdlog_logger)) {
+      setupLogger();
+    }
+
+    virtual LoggerPtr getChild(std::string tag, LoggerThreadSafety ts) = 0;
+
+   protected:
+    LoggerPtr getChildSingleTread(std::string tag, LoggerThreadSafety ts) {
+      std::string child_full_tag = tag_ + kTagHierarchySeparator + tag;
+      auto &children_by_tag = children_[tag];
+      const auto child_it = children_by_tag.find(ts);
+      if (child_it != children_by_tag.end()) {
+        return child_it->second;
+      }
+      // if a node for this child is not found in the tree config, create a
+      // new standalone logger using this logger's settings
+      auto child_config = tree_->getChild(tag).value_or(
+          std::make_shared<LoggerConfigTreeNode>(tag, tree_->getConfig()));
+      Logger new_child(
+          tag_ + kTagHierarchySeparator + tag, std::move(child_config), ts);
+      return children_by_tag
+          .emplace(std::make_pair(ts, std::make_shared<Logger>(new_child)))
+          .first->second;
+    }
+
+   private:
+    void setupLogger() {
+      static const std::map<LogLevel, const spdlog::level::level_enum>
+          kSpdLogLevels = {{LogLevel::kTrace, spdlog::level::trace},
+                           {LogLevel::kDebug, spdlog::level::debug},
+                           {LogLevel::kInfo, spdlog::level::info},
+                           {LogLevel::kWarn, spdlog::level::warn},
+                           {LogLevel::kError, spdlog::level::err},
+                           {LogLevel::kCritical, spdlog::level::critical}};
+      const auto it = kSpdLogLevels.find(tree_->getConfig().log_level);
+      BOOST_ASSERT_MSG(it != kSpdLogLevels.end(), "Unknown log level!");
+      logger_->set_level(it == kSpdLogLevels.end()
+                             ? kSpdLogLevels.at(kDefaultLogLevel)
+                             : it->second);
+      logger_->set_pattern(
+          kDefaultLogPatterns.getPattern(tree_->getConfig().log_level));
+    }
+
+    // --- Fields accessible by logger::Logger ---
+   public:
+    const std::string tag_;
+    const ConstLoggerConfigTreeNodePtr tree_;
+    const std::shared_ptr<spdlog::logger> logger_;
+
+    // --- Internal use fields ---
+   private:
+    std::unordered_map<std::string, std::map<LoggerThreadSafety, LoggerPtr>>
+        children_;
+  };
+
+  class LoggerImplSingleThread : public Logger::Impl {
```

Comment:
Seems, that this type of logger is not required now for several reasons:
- now it can be introduced only in the initialization phase, where the profit of absence of locks is not significant
- other code transformations are not assumed where it can be helpful

We discussed those arguments with the author but didn't conclude for the changes here. I request @lebdron for facilitating and help with this issue.

Reply:
thread safety option removed in f82275d